### DISCUSSION TITLE
Template Literal Types support

### DISCUSSION THREAD

**Discussion by johnsoncodehk** [vuejs/rfcs#248] [General]
TypeScript 4.1 supports more string type variants, please see: https://github.com/microsoft/TypeScript/pull/40336

It can be used to improve the type inference of props / emits... to help IDE typecheck. Backward compatibility is a problem, but we can consider use to `defineComponent` in future.

[Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBArgdgYwIZgJbCQG1QL2gXigFVEV0tcIAeAcgDEB7BgISQCdWAPGgPigHp+UGgDMGHdqxw0AUKEhQAFuEUQ4SYASgAJFWo3V4yNBmx4+g4WLYBaAEbt7SaXPDRkAWwhmtAYSRePlTKYKrqmhZCouKsEi7y7mSmlFCE-iYUeFSe3pSRwowskkjcMq4KuqH6mlQAynyEtVAQnJpwACYAzlAABgAkAN6ocCIQbFAAKgC+g8Oj40RTPVAA-JPNrWpdxGCQbMid1BN8az02gwAyDADuYwdHPDMDlWEGVESPywBcvYPTgy9qtQPksoD9agBucrQfyBSh1BpQJotNrbfpDEZjSZTc4Y+bEUGnP5PWG5LLpchBD6fMFIqFAA)
```ts
type uncapitalize = Uncapitalize<'FooBarBax'> // 'forBarBaz'
type hyphenate = Hyphenate<uncapitalize> // 'for-bar-baz'
type camelize = Camelize<hyphenate> // 'forBarBaz'
type capitalize = Capitalize<camelize> // 'FooBarBax'

type Hyphenate<S> = S extends `${infer T}${infer U}` ? T extends Uppercase<T> ? `-${Lowercase<T>}${Hyphenate<U>}` : `${T}${Hyphenate<U>}` : S;
type Camelize<S> = S extends `${infer T}-${infer U}` ? `${T}${Camelize<Capitalize<U>>}` : S;
```

**Comment by znck** [comment]
cc @pikax

**Comment by pikax** [comment]
Some work already wok the works to add for emit and events, i can add for props too 

https://github.com/vuejs/vue-next/pull/2164

**Comment by johnsoncodehk** [comment]
The closure has been implemented.
